flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBaseTest.java

```diff
@@ -0,0 +1,352 @@
+            Assert.assertEquals(1, openedTestCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, openedTestCassandraOutputFormat.getAcquiredPermits());
+
+            CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+            openedTestCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+            openedTestCassandraOutputFormat.writeRecord("N/A");
+
+            Assert.assertEquals(0, openedTestCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(1, openedTestCassandraOutputFormat.getAcquiredPermits());
+
+            // start writing
+            completableFuture.complete(null);
+
+            Assert.assertEquals(1, openedTestCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, openedTestCassandraOutputFormat.getAcquiredPermits());
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testReleaseOnFailure() throws Exception {
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            Assert.assertEquals(1, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+
+            CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+            testCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+            testCassandraOutputFormat.writeRecord("N/A");
+
+            Assert.assertEquals(0, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(1, testCassandraOutputFormat.getAcquiredPermits());
+
+            completableFuture.completeExceptionally(new RuntimeException());
+
+            Assert.assertEquals(1, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+        } catch (IOException ignored) {
+            // format.close() throws the exception gathered in format.writeRecord()
+        }
+    }
+
+    @Test(timeout = DEFAULT_TEST_TIMEOUT)
+    public void testReleaseOnThrowingSend() throws Exception {
+        Function<String, ListenableFuture<ResultSet>> failingSendFunction =
+                ignoredMessage -> {
+                    throwCheckedAsUnchecked(new Throwable("expected"));
+                    //noinspection ReturnOfNull
+                    return null;
+                };
+
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedMockOutputFormat(failingSendFunction)) {
+            Assert.assertEquals(1, testCassandraOutputFormat.getAvailablePermits());
+            Assert.assertEquals(0, testCassandraOutputFormat.getAcquiredPermits());
+
+            //noinspection OverlyBroadCatchBlock,NestedTryStatement
+            try {
+                testCassandraOutputFormat.writeRecord("none");
```

Comment:
This should fail if no exception is thrown.

Reply:
:+1: